File: JavaScript/0088. Merge Sorted Array.js
// Given two sorted integer arrays nums1 and nums2, merge nums2 into nums1 as one sorted array.
//
// Note:
//
// The number of elements initialized in nums1 and nums2 are m and n respectively.
// You may assume that nums1 has enough space (size that is greater or equal to m + n) to hold additional elements from nums2.
//
// Example:
//
// Input:
// nums1 = [1,2,3,0,0,0], m = 3
// nums2 = [2,5,6],       n = 3
//
// Output: [1,2,2,3,5,6]

/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */

/** 1) */
// Time O(n)
function merge1(nums1, m, nums2, n) {
  let i = m + n;
  m--;
  n--;

  while (i--) {
    const n1 = nums1[m];
    const n2 = nums2[n];

    if (n1 != null && n2 != null) {
      if (n1 > n2) {
        nums1[i] = n1;
        m--;
      } else {
        nums1[i] = n2;
        n--;
      }
    } else {
      if (n1 != null) {
        nums1[i] = n1;
        m--;
      } else {
        nums1[i] = n2;
        n--;
      }
    }
  }
}

/** 2) */
// Time O(n)
function merge(nums1, m, nums2, n) {
  let i = m + n;
  m--;
  n--;

  while (i--) {
    if (n < 0 || nums1[m] > nums2[n]) { // n < 0 is for the case that all nums2 has been merged, merging the rest of nums1
      nums1[i] = nums1[m--];
    } else {
      nums1[i] = nums2[n--];
    }
  }
}

File: Python/0088. Merge Sorted Array.py
# You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers m and n, representing the number of elements in nums1 and nums2 respectively.
# Merge nums1 and nums2 into a single array sorted in non-decreasing order.
# The final sorted array should not be returned by the function, but instead be stored inside the array nums1. To accommodate this, nums1 has a length of m + n, where the first m elements denote the elements that should be merged, and the last n elements are set to 0 and should be ignored. nums2 has a length of n.
#
# Example 1:
#
# Input: nums1 = [1,2,3,0,0,0], m = 3, nums2 = [2,5,6], n = 3
# Output: [1,2,2,3,5,6]
# Explanation: The arrays we are merging are [1,2,3] and [2,5,6].
# The result of the merge is [1,2,2,3,5,6] with the underlined elements coming from nums1.
#
# Example 2:
#
# Input: nums1 = [1], m = 1, nums2 = [], n = 0
# Output: [1]
# Explanation: The arrays we are merging are [1] and [].
# The result of the merge is [1].
#
# Example 3:
#
# Input: nums1 = [0], m = 0, nums2 = [1], n = 1
# Output: [1]
# Explanation: The arrays we are merging are [] and [1].
# The result of the merge is [1].
# Note that because m = 0, there are no elements in nums1. The 0 is only there to ensure the merge result can fit in nums1.
#
# Constraints:
#
# nums1.length == m + n
# nums2.length == n
# 0 <= m, n <= 200
# 1 <= m + n <= 200
# -10^9 <= nums1[i], nums2[j] <= 10^9


"""
Do not return anything, modify nums1 in-place instead.
"""


# 1) Cheating
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        nums1[m:] = nums2[:n]
        nums1.sort()


# 2)
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        i = m + n
        m -= 1
        n -= 1
        while i > 0:
            i -= 1
            n1 = nums1[m] if m >= 0 else None
            n2 = nums2[n] if n >= 0 else None
            if n1 is not None and n2 is not None:
                if n1 > n2:
                    nums1[i] = n1
                    m -= 1
                else:
                    nums1[i] = n2
                    n -= 1
            elif n1 is not None:
                nums1[i] = n1
                m -= 1
            else:
                nums1[i] = n2
                n -= 1
